File: lib/src/data/http_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

enum HttpMethod { get, post, put, delete }

typedef RequestBody = Map<String, dynamic> | List<Map<String, dynamic>> | String;

class Request {
  final String url;
  final HttpMethod method;
  final Map<String, String> headers;
  final RequestBody? body;

  const Request({
    required this.url,
    required this.method,
    required this.headers,
    this.body,
  });
}

class RequestConfig {
  final Map<String, String>? headers;
  final Future<void> Function(Request)? beforeRequest;
  final Future<void> Function(http.Response)? afterRequest;

  const RequestConfig({
    this.headers,
    this.beforeRequest,
    this.afterRequest,
  });
}

class HttpClient {
  final String baseUrl;
  Map<String, String> defaultHeaders;
  Future<void> Function(Request)? globalBeforeRequest;
  Future<void> Function(http.Response)? globalAfterRequest;

  HttpClient(this.baseUrl, [Map<String, String> defaultHeaders = const {}])
      : defaultHeaders = Map<String, String>.from(defaultHeaders) {
    // Remove trailing slash from baseUrl if present
    if (baseUrl.endsWith('/')) {
      this.baseUrl = baseUrl.substring(0, baseUrl.length - 1);
    } else {
      this.baseUrl = baseUrl;
    }
  }

  void setBeforeRequestCallback(Future<void> Function(Request)? beforeRequest) {
    globalBeforeRequest = beforeRequest;
  }

  void setAfterRequestCallback(Future<void> Function(http.Response)? afterRequest) {
    globalAfterRequest = afterRequest;
  }

  void setDefaultHeaders(Map<String, String> headers) {
    defaultHeaders = Map<String, String>.from(headers);
  }

  Future<T> request<T>(
    String endpoint,
    HttpMethod method, [
    RequestBody? body,
    RequestConfig config = const RequestConfig(),
  ]) async {
    final url = '$baseUrl$endpoint';
    final headers = {
      ...defaultHeaders,
      ...?config.headers,
    };

    String? encodedBody;
    if (body != null && body is Map<String, dynamic> || body is List<Map<String, dynamic>>) {
      encodedBody = jsonEncode(body);
    } else if (body is String) {
      encodedBody = body;
    }

    final request = Request(
      url: url,
      method: method,
      headers: headers,
      body: encodedBody,
    );

    if (globalBeforeRequest != null) {
      await globalBeforeRequest!(request);
    }

    if (config.beforeRequest != null) {
      await config.beforeRequest!(request);
    }

    try {
      final response = await _sendRequest(request);

      if (globalAfterRequest != null) {
        await globalAfterRequest!(response);
      }

      if (config.afterRequest != null) {
        await config.afterRequest!(response);
      }

      final responseBody = await _parseResponse<T>(response);
      return responseBody;
    } catch (error) {
      if (error is Exception) {
        rethrow;
      }
      throw Exception(error.toString());
    }
  }

  Future<http.Response> _sendRequest(Request request) async {
    switch (request.method) {
      case HttpMethod.get:
        return await http.get(
          Uri.parse(request.url),
          headers: request.headers,
        );
      case HttpMethod.post:
        return await http.post(
          Uri.parse(request.url),
          headers: request.headers,
          body: request.body,
        );
      case HttpMethod.put:
        return await http.put(
          Uri.parse(request.url),
          headers: request.headers,
          body: request.body,
        );
      case HttpMethod.delete:
        return await http.delete(
          Uri.parse(request.url),
          headers: request.headers,
          body: request.body,
        );
    }
  }

  Future<T> _parseResponse<T>(http.Response response) async {
    if (response.body.isEmpty) {
      return {} as T;
    }

    try {
      final decodedBody = jsonDecode(response.body);
      return decodedBody as T;
    } catch (e) {
      // If JSON parsing fails, return the raw body as string
      return response.body as T;
    }
  }

  Future<T> get<T>(
    String endpoint, [
    RequestBody? body,
    RequestConfig? config,
  ]) async {
    return request<T>(endpoint, HttpMethod.get, body, config ?? const RequestConfig());
  }

  Future<T> post<T>(
    String endpoint, [
    RequestBody? body,
    RequestConfig? config,
  ]) async {
    return request<T>(endpoint, HttpMethod.post, body, config ?? const RequestConfig());
  }

  Future<T> put<T>(
    String endpoint,
    RequestBody body, [
    RequestConfig? config,
  ]) async {
    return request<T>(endpoint, HttpMethod.put, body, config ?? const RequestConfig());
  }

  Future<T> delete<T>(
    String endpoint, [
    RequestConfig? config,
  ]) async {
    return request<T>(endpoint, HttpMethod.delete, null, config ?? const RequestConfig());
  }
}

File: src/HttpClient.ts
type HttpMethod = "GET" | "POST" | "PUT" | "DELETE";

type Body = Record<string, any> | Record<string, any>[] | string;

type Request = {
  url: string
  method: HttpMethod
  headers: Record<string, string>,
  body?: BodyInit
}

interface RequestConfig {
  headers?: Record<string, string>;
}

export class HttpClient {
  private baseUrl: string;
  private defaultHeaders: Record<string, string>;
  private globalBeforeRequest?: (
    req: Request
  ) => void | Promise<void>;
  private globalAfterRequest?: (response: Response, responseBody: any) => void | Promise<void>;

  constructor(baseUrl: string, defaultHeaders: Record<string, string> = {}) {
    this.baseUrl = baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
    this.defaultHeaders = defaultHeaders;
  }

  setBeforeRequestCallback(
    beforeRequest?: (req: Request) => void | Promise<void>,
  ) {
    this.globalBeforeRequest = beforeRequest;
  }

  setAfterRequestCallback(
    afterRequest?: (response: Response) => void | Promise<void>,
  ) {
    this.globalAfterRequest = afterRequest;
  }

  setDefaultHeaders(headers: Record<string, string>) {
    this.defaultHeaders = headers;
  }

  private async request<T>(
    endpoint: string,
    method: HttpMethod,
    body?: Body,
    config: RequestConfig = {},
  ): Promise<T> {
    const url = `${this.baseUrl}${endpoint}`;
    const headers = {
      ...this.defaultHeaders,
      ...config.headers,
    };

    if (body && typeof body === "object") {
      body = JSON.stringify(body);
    }
    
    const request: Request = {
      url,
      headers,
      method,
      body
    }
    if (this.globalBeforeRequest) {
      await this.globalBeforeRequest(request);
    }
    
    try {
      const response = await fetch(request.url, {
        method: request.method,
        headers: request.headers,
        body: request.body,
      });

      const responseBody = await response.json().catch(e => undefined);

      if (this.globalAfterRequest) {
        await this.globalAfterRequest(response, responseBody);
      }

      return responseBody;
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error(String(error));
    }
  }

  async get<T>(endpoint: string, body?: Body, config?: RequestConfig): Promise<T> {
    return this.request(endpoint, "GET", body, config) as T;
  }

  async post<T>(
    endpoint: string,
    body?: Body,
    config?: RequestConfig,
  ): Promise<T> {
    return this.request(endpoint, "POST", body, config) as T;
  }

  async put<T>(
    endpoint: string,
    body: Body,
    config?: RequestConfig,
  ): Promise<T> {
    return this.request(endpoint, "PUT", body, config) as T;
  }

  async delete<T>(endpoint: string, config?: RequestConfig): Promise<T> {
    return this.request(endpoint, "DELETE", undefined, config) as T;
  }
}
